### PR TITLE
Post demystifying `IndexByType`

### PR DESCRIPTION
**Pull request by Andarist**
This is a strong draft of the post - it needs to be polished before we can publish it.

I think it has the potential of splitting into two parts:
1. typegen intro: outline the problem, describe the goals (stuff up to the "indexing" header)
2. `IndexByType` demystified (the big part from the "indexing" header)

We could make this whole thing a series of blog posts about the typegen's internal machinery. Not every blog post has to be as long as this one - some tricks used in there are much simpler, or at least they can be described much quicker. And I think that each blog post should be its own thing, it's not worth squeezing two different tricks into a single post or something like that.

I would be very grateful for some initial feedback and help with the structure, language, etc

### CODE REVIEW

File: content/posts/2022-03-28-mapped-types.mdx [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+
+Let's describe what happens here in plain english: iterate through the input union (`T`), assign the current member to `E`, set the key to its type (`E["type"]`) and the value to the current member itself (`E`).
+
+What is really neat here is that we get access to non-string type that we can use to compute the value as long as we "set" the key to a string type. This was not possible before this feature was implemented. We only could iterate over the strings as the current element of the iteration was always also used as the key. So, in the past, we had to retrieve any "non-primitive" type using our current key from some other type - and this is what basically we've been doing in the previous versions of this `IndexByType` type.
+
+Note: we can also use numbers and symbols as the type for the key. I've left it out for brevity of the explanation.
+
+One small issue with this solution is that values still have those string unions as type in them. This isn't incorrect but perhaps it can be viewed as a little bit confusing. Consider the situation from before:
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:406]:
```suggestion
One small issue with this solution is that values still have those string unions as types. This isn't incorrect but perhaps could be viewed as a little confusing. Consider the situation from before:
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+        count: ctx.count + event.value,
+      };
+    }),
+  },
+}
+```
+
+In here we know that `event.type` can only be of type `"INC"` but if we provide a type for it based on the result of our `IndexByType` then it's going to be of type `"INC" | "DEC"`.
+
+As mentioned, this isn't incorrect and we don't plan to "fix" this in XState as that would potentially add unneccessary overhead to actually narrow this even further. The return of this work would also be questionable, cause if the user declares their event type like this then probably they use those types in their function signatures etc anyway. And that's the only moment where this should matter - when one pass the received `event` to some other code that expect a certain type.
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:422]:
```suggestion
As I previously mentioned, this isn’t incorrect, and we don't plan to “fix” it in XState, as narrowing this further would potentially add unnecessary overhead. The value of this work would also be questionable because if the user declares their event type in this way, they probably already use those types in their function signatures, etc. And the only moment where this should matter is when the received `event` is passed to some other code that expects a certain type
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+
+Just for fun we can think through how we could actually solve this problem. The simplest solution that we can try is this:
+
+```ts
+type IndexByType<T extends { type: string }> = {
+  [E in T as E["type"]]: E & { type: E["type"] };
+};
+```
+
+However, this doesn't work properly because the `E` on the right is already the type with a `"type"` property that is the union. After all, this is the exact type that we currently see as those values. We can't easily access the `E["type"]` from the left side (from the key part). So we have to introduce a second pass over this type.
```
[the remark this answers is not in the document]
Comment (laurakalbag) [reply content/posts/2022-03-28-mapped-types.mdx:432]: I think this sentence could be clearer, but I’m not sure how to edit it without changing its meaning.

File: content/posts/2022-03-28-mapped-types.mdx [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+type NarrowType<T> = {
+  [K in keyof T]: T[K] & { type: K };
+};
+
+type IndexByType<T extends { type: string }> = NarrowType<{
+  [E in T as E["type"]]: E;
+}>;
+```
+
+This one works but it suffers from the readability problem that we've discussed earlier:
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:444]:
```suggestion
The following solution works but suffers from the readability problem we discussed earlier:
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+  };
+  RESET: {
+    type: "RESET";
+  } & {
+    type: "RESET";
+  };
+};
+```
+
+So the solution to this might be to reuse our `Compute` helper from before within `NarrowType`:
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:468]:
```suggestion
Therefore, the solution might be to reuse our `Compute` helper from before within `NarrowType`:
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+
+```ts
+type NarrowType<T> = {
+  [K in keyof T]: Compute<T[K] & { type: K }>;
+};
+```
+
+and this returns a much more readable type:
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:476]:
```suggestion
which returns a much more readable type:
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+    type: "DEC";
+    value: number;
+  };
+  RESET: {
+    type: "RESET";
+  };
+};
+```
+
+However, what if our input union would be defined like this?
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:494]:
```suggestion
But what if our input union is defined as follows?
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+
+```ts
+interface ResetEvent {
+  type: "RESET";
+}
+type MyEvent = { type: "INC" | "DEC"; value: number } | ResetEvent;
+```
+
+With a union defined like this we would still get the same result as above. Using `Compute` always produce "anonymous types" so all that we can see is their content being "inlined". But in here we actually gave some name to one of the events - there is `ResetEvent` and it would be great if we could preserve that.
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:503]:
```suggestion
We would still get the same result as the previous example with a union defined in this way. Using `Compute` always produces *anonymous types* so all we can see is their content being *inlined*. But here, we gave a name to one of the events, `ResetEvent`. And it would be great if we could preserve the name.
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+
+So when do we actually need to use `Compute` in the first place? Only when the `T[K]["type"]` **isn't exactly** matching the `K`. Let's try the adjusted version of the `NarrowType` then:
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:505]:
```suggestion
So when do we need to use `Compute`? Only when the `T[K]["type"]` **isn’t exactly** matching the `K`. Let’s try the adjusted version of the `NarrowType` then:
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 10]
```diff
@@ -0,0 +1,532 @@
...
+
+```ts
+type NarrowType<T> = {
+  [K in keyof T]: T[K]["type"] extends K ? T[K] : Compute<T[K] & { type: K }>;
+};
+```
+
+And with that we get:
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:513]:
```suggestion
And with the adjusted version, we get:
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 11]
```diff
@@ -0,0 +1,532 @@
...
+
+- preserved names for event types that have literal type property
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:515]:
```suggestion
- preserved names for event types that have the literal type property
- narrowed down/computed event types for event types using unions for their type property
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 12]
```diff
@@ -0,0 +1,532 @@
...
+  };
+  DEC: {
+    type: "DEC";
+    value: number;
+  };
+  RESET: ResetEvent;
+};
+```
+
+It's also worth noting that our check isn't literally checking if `T[K]["type"]` is exactly the same as `K`. It's checking if it is a subtype of `K`. But we are working on a very strong, practical, assumption that each member of the union has a literal type as its type (or a union of those) and none of them is using `{ type: string }`.
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:532]:
```suggestion
It’s also worth noting our check isn’t literally checking if `T[K]["type"]` is exactly the same as `K`. Instead, it’s checking if `T[K]["type"]` is a subtype of `K`. But we’re working on a very strong practical assumption that each member of the union has a literal type, as its type (or a union of those) and none of them is using `{ type: string }`
```

Comment (laurakalbag) [reply content/posts/2022-03-28-mapped-types.mdx:532]: Perhaps we could finish off with a short conclusion/summary?

File: content/posts/2022-03-28-mapped-types.mdx [hunk 13]
```diff
@@ -0,0 +1,532 @@
+---
+title: mapped types
+description:
+keywords:
+  - typescript
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:5]:
```suggestion
  - typescript
  - xstate
  - typegen
```

File: content/posts/2022-03-28-mapped-types.mdx [hunk 14]
```diff
@@ -0,0 +1,532 @@
+---
+title: mapped types
```
Comment (laurakalbag) [inline content/posts/2022-03-28-mapped-types.mdx:2]:
```suggestion
title: Demystifying IndexByType using mapped types
```

**Review by laurakalbag — CHANGES_REQUESTED** [review]
I think the structure is fine as a long read like this will really appeal to folks who like to know the details, which is really the point of the post!

A few more headings might help give some navigational landmarks which will also appear in the Table of contents sidebar. I’ve noted where I think we might need them, but you know best.

Most of my suggestions are minor grammatical fixes, some typography and occasionally trying to make it clearer what you’re referring to when you say “it”, “this”, or “that.” Please feel free to dismiss or ignore any you disagree with, and let me know if you want a re-review at any point.

Let me know when you’ve confirmed the title and I’ll make a matching open graph image so it looks pretty on social media.

File: content/posts/2022-03-28-mapped-types.mdx (hunk 12 above)
Comment (Andarist) [reply content/posts/2022-03-28-mapped-types.mdx:532]: Ye, I was thinking about adding some closing thoughts. What do you think about "please don't do it at home"? 🤣

Comment (laurakalbag) [reply content/posts/2022-03-28-mapped-types.mdx:532]: “I did this so you don’t have to” 😂

Comment (Andarist) [reply content/posts/2022-03-28-mapped-types.mdx:532]: that's a good one!

File: content/posts/2022-03-28-mapped-types.mdx [hunk 15]
```diff
@@ -0,0 +1,540 @@
...
+  DEC: DecrementEvent;
+  RESET: ResetEvent;
+};
+```
+
+There is no real difference between those two results; they should be functionally equivalent. However, for readability purposes, you might prefer to use the *computed variant*, as the string representation of this type might end up printed in TS tooltips when we make a mistake or simply hover over some type that refers to it.
+
+Honestly, I’ve not reached for the showcased implementation so far. I’ve just shared it as a comparison to the “JavaScript implementation” and for fun 😉
+
+In XState right now, we are using the following implementation:
```
Comment (Andarist) [inline content/posts/2022-03-28-mapped-types.mdx:275]: Laura has suggested adding a subheading here: https://github.com/statelyai/eng-blog/pull/81#discussion_r838408731
